### ISSUE TITLE
`cmake` not defined

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#620]
Spack defines a `cmake` command by default. I find that this is `None` unless `cmake` is installed on the system; Spack does not pick up `cmake` as installed by a dependency on `cmake.`

I'm looking at `setup_package` -- maybe `which("cmake")` is called before the dependencies have a chance to modify `PATH`?

**Comment by eschnett** [comment]
I forgot to add: If I add `cmake = which('cmake')` in the respective package, all works fine.

**Comment by tgamblin** [comment]
Yep it's an ordering thing in `build_environment.py`.  Will send a fix momentarily.

**Comment by citibeth** [comment]
I'm dubious about all these "convenience" things that get added to the environment:

1. They're non-standard Python, so I can never remember how they work.  I understand "self.".
2. They're (temporarily) part of the the module not the class.  Although I can't say if or when this might hurt us for the particular usage in Spack, I know this is not clean OO programming.
3. They're not always available.  In my StagedPackage class, they're NOT available, so I used the "native" Python methods (via self.) instead.  Same thing in git/package.py, I had to write "which('autoconf')('-i')".  Having two ways to do it is harder than having just one.

**Comment by tgamblin** [comment]
I'm becoming more dubious about them too.

I think they should probably go into two places eventually:

  1. The `spack` namespace, or some other namespace that can be easily imported into package files.  `from spack import *` was intended to be simple and to provide this stuff.
  2. Make them attributes on the spec (e.g., so that deps can make fields available to a package).  Some packages like `netlib-scalapack` already do this, so you can write:

      spec['scalapack'].fc_link

    in packages that depend on `scalapack`.

I'm not ready to deprecate this stuff yet though -- it's still used and I'd like to see how the packages evolve before throwing yet another not-quite-good-enough convention in there.

**Comment by eschnett** [comment]
@tgamblin I still see this problem on the current master. Since this is a show-stopper, can we quickly add the respective `cmake = which('cmake')` calls to all packages? We can easily revert that patch later.

**Comment by citibeth** [comment]
PR #664 patches cmake --> which('cmake') for all packages.

**Comment by eschnett** [comment]
@citibeth Thanks; I read about this. That's essentially what I did on my master branch as well.

Maybe the Spack self-tests should include installing some small packages, on a bare-bones Ubuntu virtual machine?

**Comment by davydden** [comment]
> Maybe the Spack self-tests should include installing some small packages, on a bare-bones Ubuntu virtual machine?

and OSX perhaps?

A bit out-of-topic, but one could think about using Travis to build packages, changed in PRs at least on one system (e.g. Ubuntu 14.04 LTS?). Just to be able to check that nothing is broken at least there with GNU compilers. Ideally one would also build on OSX and maybe another Linux with Intel compilers to try to cover bigger ground.

**Comment by eschnett** [comment]
Ping: Please either correct this in Spack, or apply @citibeth's patch.

**Comment by goxberry** [comment]
AFAICT, `cmake@3.5.1` is still broken; not sure if this error is related to the issues in this thread. Log of `spack install --verbose cmake` is below:

```
narella:spack oxberry1$ bin/spack install --verbose cmake
==> Installing cmake
==> ncurses is already installed in /Users/oxberry1/spack/opt/spack/darwin-x86_64/gcc-5.3.0/ncurses-6.0-f2klgir4qnijs3fensv7ysloknqas7hi
==> openssl is already installed in /Users/oxberry1/spack/opt/spack/darwin-x86_64/gcc-5.3.0/openssl-1.0.2g-zi52d5dlje7ipf4uprnu7u2kgtcauhwi
==> Already downloaded /Users/oxberry1/spack/var/spack/stage/cmake-3.5.1-tgxyhlo2ckyde7erzn4chgq7kye3axui/cmake-3.5.1.tar.gz
==> Already staged cmake-3.5.1-tgxyhlo2ckyde7erzn4chgq7kye3axui in /Users/oxberry1/spack/var/spack/stage/cmake-3.5.1-tgxyhlo2ckyde7erzn4chgq7kye3axui
==> No patches needed for cmake
==> Building cmake
==> ./configure --prefix=/Users/oxberry1/spack/opt/spack/darwin-x86_64/gcc-5.3.0/cmake-3.5.1-tgxyhlo2ckyde7erzn4chgq7kye3axui --parallel=4 -- -DCMAKE_USE_OPENSSL=ON
---------------------------------------------
CMake 3.5.1, Copyright 2000-2016 Kitware, Inc.
C compiler on this system is: /Users/oxberry1/spack/lib/spack/env/gcc/gcc
C++ compiler on this system is: /Users/oxberry1/spack/lib/spack/env/gcc/g++
Makefile processor on this system is: make
/Users/oxberry1/spack/lib/spack/env/gcc/g++ is GNU compiler
/Users/oxberry1/spack/lib/spack/env/gcc/g++ has setenv
/Users/oxberry1/spack/lib/spack/env/gcc/g++ has unsetenv
/Users/oxberry1/spack/lib/spack/env/gcc/g++ does not have environ in stdlib.h
/Users/oxberry1/spack/lib/spack/env/gcc/g++ has stl wstring
---------------------------------------------
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmBootstrapCommands1.cxx -o cmBootstrapCommands1.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmCustomCommandGenerator.cxx -o cmCustomCommandGenerator.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmCacheManager.cxx -o cmCacheManager.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmListFileCache.cxx -o cmListFileCache.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmComputeLinkDepends.cxx -o cmComputeLinkDepends.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmComputeLinkInformation.cxx -o cmComputeLinkInformation.o
/Users/oxberry1/spack/lib/spack/env/gcc/g++  -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk -I/private/var/folders/37/zqjg7vn9
7kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source   -I/private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Utiliti
es -c /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmOrderDirectories.cxx -o cmOrderDirectories.o
In file included from /usr/include/dispatch/dispatch.h:51:0,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:15,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:13,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:55,
                 from /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmFindProgramCommand.cxx:16,
                 from /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Source/cmBootstrapCommands1.cxx:52:
/usr/include/dispatch/object.h:143:15: error: expected unqualified-id before '^' token
 typedef void (^dispatch_block_t)(void);
               ^
/usr/include/dispatch/object.h:143:15: error: expected ')' before '^' token
/usr/include/dispatch/object.h:362:3: error: 'dispatch_block_t' has not been declared
   dispatch_block_t notification_block);
   ^
make: *** [cmBootstrapCommands1.o] Error 1
make: *** Waiting for unfinished jobs....
---------------------------------------------
Error when bootstrapping CMake:
Problem while running make
---------------------------------------------
Log of errors: /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/Bootstrap.cmk/cmake_bootstrap.log
---------------------------------------------
==> Error: Command exited with status 9:
./configure --prefix=/Users/oxberry1/spack/opt/spack/darwin-x86_64/gcc-5.3.0/cmake-3.5.1-tgxyhlo2ckyde7erzn4chgq7kye3axui --parallel=4 -- -DCMAKE_USE_OPENSSL=ON

See build log for details:
  /private/var/folders/37/zqjg7vn97kggmzjdbcm55j6h001b24/T/oxberry1/spack-stage/spack-stage-Zs9eGR/cmake-3.5.1/spack-build.out

/Users/oxberry1/spack/var/spack/repos/builtin/packages/cmake/package.py:87, in install:
     68       def install(self, spec, prefix):
     69           # Consistency check
     70           self.validate(spec)
     71
     72           # configure, build, install:
     73           options = ['--prefix=%s' % prefix]
     74           options.append('--parallel=%s' % str(make_jobs))
     75
     76           if '+qt' in spec:
     77               options.append('--qt-gui')
     78
     79           if '+doc' in spec:
     80               options.append('--sphinx-html')
     81               options.append('--sphinx-man')
     82
     83           if '+openssl' in spec:
     84               options.append('--')
     85               options.append('-DCMAKE_USE_OPENSSL=ON')
     86
  >> 87           configure(*options)
     88           make()
     89           make('install')
==> Error: Installation process had nonzero exit code.
narella:spack oxberry1$ git rev-parse HEAD
8ef9d685427e60ce5af660be5f27d36683ea8845
```

**Comment by tgamblin** [comment]
@goxberry: This looks like a CMake build failure (not that the `cmake` symbol isn't properly in scope).  Does `cmake` work when you build it with clang on OS X?  I think the issue is that it's not building properly with gcc 5.  This works for me:

```
    spack install cmake@3.5.1 %clang
```

I think this could be solved with some better defaults so that `cmake` can prefer a compiler when one isn't specified -- there isn't a reason that packages like this *need* to build with the same compiler as the code -- adding this is on my list after #378.  

For now, does this work?

```
    spack install metis %gcc ^cmake %clang
```

**Comment by goxberry** [comment]
@tgamblin I think I responded to this on #749, but for completeness, `spack install metis %gcc ^cmake clang` does work. The tricky thing about this fix is that if a spack package depends on a package that depends on cmake -- that is, cmake is a second- (or higher-) order dependency -- then this command won't work, because spack returns an error saying that `<package> does not depend on cmake`. (Which is sort of accurate -- cmake is not a direct (or first-order) dependency.)

**Comment by mathstuf** [comment]
Yeah, that's a problem with Xcode changing headers and GCC not being compatible with some Objective-C code in it. Should have been fixed in 3.5.1 by cmake/cmake@d84ba668d7dd77ac38b3795f92e4ef0f894c696e though. It's a separate issue anyways.

**Comment by adamjstewart** [comment]
I just tried installing a package that builds with cmake but doesn't depend on cmake. It crashed because I'm on my laptop and I don't have cmake installed. When I added `depends_on('cmake')`, Spack built cmake and correctly used it to build the package. I think the scope thing must have gotten figured out. Should be able to close this issue now.

**Comment by manish2479** [comment]
Hi 

I am not able to install cmake with qt. Here is the error I am getting 

==> './bootstrap' '--prefix=/home/user1/spack/opt/spack/linux-sles12-x86_64/gcc-4.8/cmake-3.8.1-w4cp2rscxv562fo4ez54awgto5f7plss' '--parallel=32' '--no-system-jsoncpp' '--system-libs' '--qt-gui' '--' '-DCMAKE_USE_OPENSSL=ON'
---------------------------------------------
CMake 3.8.1, Copyright 2000-2017 Kitware, Inc. and Contributors
C compiler on this system is: /home/user1/spack/lib/spack/env/gcc/gcc
C++ compiler on this system is: /home/user1/spack/lib/spack/env/gcc/g++
Makefile processor on this system is: gmake
/home/user1/spack/lib/spack/env/gcc/g++ is GNU compiler
/home/user1/spack/lib/spack/env/gcc/g++ has setenv
/home/user1/spack/lib/spack/env/gcc/g++ has unsetenv
/home/user1/spack/lib/spack/env/gcc/g++ does not have environ in stdlib.h
/home/user1/spack/lib/spack/env/gcc/g++ has stl wstring
/home/user1/spack/lib/spack/env/gcc/g++ has <ext/stdio_filebuf.h>
---------------------------------------------
gmake: 'cmake' is up to date.
loading initial cache file /tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Bootstrap.cmk/InitialCacheFlags.cmake
-- Using system-installed BZIP2
-- Using system-installed CURL
-- Using system-installed EXPAT
-- Using system-installed FORM
-- Using system-installed LIBARCHIVE
-- Using system-installed LIBLZMA
-- Using system-installed LIBRHASH
-- Using system-installed LIBUV
-- Using system-installed ZLIB
-- Could NOT find LibRHash (missing:  LibRHash_LIBRARY LibRHash_INCLUDE_DIR)
CMake Error at CMakeLists.txt:309 (message):
  CMAKE_USE_SYSTEM_LIBRHASH is ON but LibRHash is not found!
Call Stack (most recent call first):
  CMakeLists.txt:663 (CMAKE_BUILD_UTILITIES)


-- Configuring incomplete, errors occurred!
See also "/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeOutput.log".
:

Determining if compiler supports C11 _Thread_local failed with the following output:
Change Dir: /tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_5620d/fast"
gmake -f CMakeFiles/cmTC_5620d.dir/build.make CMakeFiles/cmTC_5620d.dir/build
gmake[1]: Entering directory '/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_5620d.dir/cm_c11_thread_local.c.o
/home1/user1/spack/lib/spack/env/gcc/gcc   -std=gnu11 -o CMakeFiles/cmTC_5620d.dir/cm_c11_thread_local.c.o   -c /tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Source/Checks/cm_c11_thread_local.c
/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Source/Checks/cm_c11_thread_local.c:1:15: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'int'
 _Thread_local int i = 42;
               ^
CMakeFiles/cmTC_5620d.dir/build.make:65: recipe for target 'CMakeFiles/cmTC_5620d.dir/cm_c11_thread_local.c.o' failed
gmake[1]: *** [CMakeFiles/cmTC_5620d.dir/cm_c11_thread_local.c.o] Error 1
gmake[1]: Leaving directory '/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeTmp'
Makefile:126: recipe for target 'cmTC_5620d/fast' failed
gmake: *** [cmTC_5620d/fast] Error 2


Determining if compiler supports C++ auto_ptr failed with the following output:
Change Dir: /tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_5af52/fast"
gmake -f CMakeFiles/cmTC_5af52.dir/build.make CMakeFiles/cmTC_5af52.dir/build
gmake[1]: Entering directory '/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_5af52.dir/cm_cxx_auto_ptr.cxx.o
/home1/user1/spack/lib/spack/env/gcc/g++    -std=gnu++1y -o CMakeFiles/cmTC_5af52.dir/cm_cxx_auto_ptr.cxx.o -c /tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Source/Checks/cm_cxx_auto_ptr.cxx
/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Source/Checks/cm_cxx_auto_ptr.cxx: In function 'std::auto_ptr<int> get_auto_ptr()':
/tmp/user1/spack-stage/spack-stage-Qn2BTX/cmake-3.8.1/Source/Checks/cm_cxx_auto_ptr.cxx:5:22: warning: 'auto_ptr' is deprecated (declared at /usr/include/c++/4.8/backward/auto_ptr.h:87) [-Wdeprecated-declarations]
   std::auto_ptr<int> ptr;

Any help would be appreciated.

Thank You

**Comment by adamjstewart** [comment]
Sounds like the package is missing a dependency on LibRHash. @mathstuf Can you confirm?

**Comment by mathstuf** [comment]
Yes, `librhash` is a dependency as of 3.8.0.

**Comment by adamjstewart** [comment]
This should be fixed by #5120. @manish2479 can you test this out?
